homeassistant/components/tolo/sensor.py

```diff
@@ -15,6 +22,75 @@
 from .const import DOMAIN
 
 
+@dataclass
+class ToloSensorEntityDescriptionBase:
+    """Required values when describing TOLO Sensor entities."""
+
+    getter: Callable[[StatusInfo], int | None]
+    availability_checker: Callable[[SettingsInfo, StatusInfo], bool] | None
+
+
+@dataclass
+class ToloSensorEntityDescription(
+    SensorEntityDescription, ToloSensorEntityDescriptionBase
+):
+    """Class describing TOLO Sensor entities."""
+
+    state_class = SensorStateClass.MEASUREMENT
+
+
+SENSORS = (
+    ToloSensorEntityDescription(
+        key="water_level",
+        entity_category=EntityCategory.DIAGNOSTIC,
+        icon="mdi:waves-arrow-up",
+        name="Water Level",
+        native_unit_of_measurement=PERCENTAGE,
+        getter=lambda status: status.water_level_percent,
+        availability_checker=None,
+    ),
+    ToloSensorEntityDescription(
+        key="tank_temperature",
+        device_class=SensorDeviceClass.TEMPERATURE,
+        entity_category=EntityCategory.DIAGNOSTIC,
+        name="Tank Temperature",
+        native_unit_of_measurement=TEMP_CELSIUS,
+        getter=lambda status: status.tank_temperature,
+        availability_checker=None,
+    ),
+    ToloSensorEntityDescription(
+        key="power_timer_remaining",
```

Comment:
Countdowns need to have the `SensorDeviceClass.TIMESTAMP` device class and provide a timestamp of when the countdown finishes.

Reply:
Mh... how to deal with the issue that I don't know seconds of the minute-based timers and the minutes and seconds of the hour-based timer? The device just gives me "in 5 minutes" and will change that information after one minute. If I do some datetime math and e.g. just add the minutes this would result in a value jumping back and forth:

Let's assumed the timer was started at 20:15:25 and I just add real time + timer value:

| Real Time | Timer Value | Resulting Timestamp |
|-----|-----|-----|
| 20:15:25  | 5 | 20:20:25 |
| 20:16:05  | 5 | 20:21:05 |
| 20:16:25  | 4 | 20:20:25 |
| 20:17:05  | 4 | 20:21:05 |
| 20:17:25  | 3 | 20:20:25 |
| 20:18:05  | 3 | 20:21:05 |
| 20:18:25  | 2 | 20:20:25 |
| 20:19:05  | 2 | 20:21:05 |
| 20:19:25  | 1 | 20:20:25 |
| 20:20:05  | 1 | 20:21:05 |

This will get even worse with the hour based timer, which, set to 7 hours, will (after 59 minutes) still report 7 hours (will change to 6 after 60 minutes). Therefore, the timestamp will be wrong by ~one hour. This even happens if I e.g. always assume seconds (and minutes) to be 0 - either the assumed end timestamp is jumping or is incredibly unprecise. Any idea?

Reply:
That is actually really good reasoning, I wonder if we should make an exception for this case.

@MartinHjelmare May I ask for a second opinion on that one?

Reply:
The only other solution I can think of right now is that we don't update state again until the timer runs out, ie until current time passes the timestamp of the existing state.

Reply:
Another idea: We could store the old timer value besides storing the timestamp as state. We only update state if the new timer value changes from the old timer value, eg from 5 to 4. If we then don't include more precision in our state timestamp than what the timer value represents, eg minutes or hours, we should be able to avoid jumps in the timestamp.

Does that sound true?

Reply:
I see several problems when we start to try to fake the timestamp here:

First, we would have to consider all the edge cases (including the one mentioned by @MartinHjelmare, what happens if the timer was changed during operation). Manually decreasing the timer setting during operation is somehow easy to detect. But what happens if, e.g., after 59 minutes the hourly timer was turned of and on again (which might happen in real use). This would "reset" the hourly timer from 7:01 (internal value, sent by the device as `8`, since only 7:00 will change it to `7`) to 8:00 (sent as `8`). Our logic here could not detect that if power down/up was done faster then pull interval (5 seconds), which is quite easy to do, it's just pressing the power button twice. Then the display in HA would show the timesamp ending 59 minutes too early.

Another edge case: What happens if HA is rebooting during device hands-on actions starting or stopping or changing a timer? E.g., the physical setup I'm going to build in my house: I have a tablet in the upstairs living room, where I can start the sauna (preheat, etc...). When it's fully heated, I will go down. After finishing sauna time, I directly deactivate the sauna device from a control panel which is directly in the sauna room and part of the device (water proof, so I prefer this to have in the steam cabine over a, e.g., Android tablet). Changes I do on this panel should correctly be visible to HA, e.g. if I go upstairs I want to check if everything (fan) turned off after a while, when the sauna is assumed to be cold and dry again. Same for the hot water tank: It looks crazy if the one sensor (water tank level) shows "empty", while the timer sais "I will empty that in 31 minutes and 45 seconds (this is part of the meaning of the hourly sweep timer).

I'm sure there are pretty more edge cases which we didn't discuss or think about yet. Furthermore, testing is quite heavy with an hour based timer and a device with 7.5kW power consumption... :laughing:  So figuring out bugs could take a long time which I then would better need for paid work for the incoming energy bill. :laughing: 

Currently, the whole `tolo` integration (regarding device communication) is completely stateless. The device's state is reflected in 25 bytes (17 status, 8 settings, see https://gitlab.com/MatthiasLohr/tololib/-/blob/main/tololib/state.py) which is received by the `tololib` dependency and after conversion to native Python types directly passed to and displayed by HA. Going the way of faking a value which is actually not there increases the total complexity of the integration and therefore increases the difficulty of maintenance, too.

I would strongly vote for an exception in this case and just displaying the original value in a sensor as it currently done in the PR. My best and favourite argument: By convention, you want to show a timestamp (as it gives best precision, which I completely understand). But faking a precision which we don't have with heavy and most probably buggy (e.g. not considering all edge cases) logic and then probably showing the user  wrong in information in high precision instead of showing the correct information in the original lower precision... it just feels quite wrong.

Reply:
You're first problem example is due to the involved precision from the API. If the API doesn't give better precision than hours the state can indeed be minutes off.

I don't follow your second paragraph. Sorry.

We want to use a timestamp as state instead of relative time when the relative time changes just because time passes. We don't want our state to change just because time passes.

I think it's ok to use a relative time, as an exception here, if we don't update state more frequently than once per minute.

Reply:
Currently, we update all states of the `tolo` integration (e.g., temperature, water tank level, water tank temperature, ...) on a 5 second basis (since temperatures can actually change quite fast), sending one status request message and getting back a 17 byte UDP package, containing all the information, including the information we're discussing here. For reasons of simplicity, I would just return this value (as already coded in the pull request). Then, theoretically, it can get updated more frequently than once per minute. To be clear: Merging the pull request as it is does not increase network traffic/polls/..., it just returns a value which is already known to the coordinator, since one status message contains. There would be a specific request/response possibilities for 3 of 4 of the possible timers, only sending back a single byte (plus overhead), but I would not use it since the information is already there in the coordinator.

Alternatively, we decrease the the local pull interval for the whole integration, which significantly decreases UX (temperature values are quite relevant for the integration and may increase mich faster than one minute can reflect and might be useful to the user). E.g., with the 5 seconds update interval, I can see if the water tank is getting refilled (completely automatically), so the temperature decreases slightly, I can see the heater starting again, bringing the temperature up to 100°, all happening within 2-15 seconds (depending on the water inbound temperature). So, for UX, in my opinion not a good choice and significally reduces usefulness of the integration.

Third option: In the getter, add new code which "caches" the last reply until one minute has passed and then use the actual value already known to the coordinator - and this does not make any sense I guess.

Reply:
> I think it's ok to use a relative time, as an exception here, if we don't update state more frequently than once per minute.

Just for understanding... *state change* is different from *returning a value for the sensor*, right? A *state change* only occurs if the value getter returns something different to its previous call, right? If yes: Forget about my previous comment, then it's already implemented that way.

Except heavy strange behavior of people administrating the device (turning it on/off all the time or changing timer values all the time), the return value of the value getter will only change once per minute.

Reply:
The state won't be updated in the state machine unless the state or state attributes change, unless the entity has set `force_update` to True.

Reply:
Then... it's all set. I don't play around with `force_update` at all and (except of strange user behaviour) the value stays the same. And if you would be willing to accept the exception for this sensors to allow int (minutes) instead of timestamp for the timer... everything is implemented.

Thanks for the discussion @MartinHjelmare and @frenck!

Personal note: Sorry if I'm annoying sometimes, I just want to get the best outcome for HA/users of my integration, but I also have to learn a lot about HA, so I ask a lot of questions to really understand things...

Reply:
>  I just want to get the best outcome for HA/users of my integration

We have a common goal 😉

Reply:
As there are no other attributes, and all units are minutes, and the native type is an integer. I think this is OK in that case, as we agreed to allow for it.

👍

Reply:
Thank you very much!